### PR TITLE
[enhancement](nereids) Speedup partition pruner

### PR DESCRIPTION
**Pull request by 924060929**
## Proposed changes

1. fast return when partition predicate is true/false/null
2. fast compute table's hash code
3. fast merge two ranges when equals

### CODE REVIEW

**Comment by 924060929** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java [hunk 1]
```diff
@@ -1989,9 +1989,7 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        return Objects.hash(state, indexIdToMeta, indexNameToId, keysType, partitionInfo, idToPartition,
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java:1992]: 1. selectedIndex should be add
2. do we need to hash all these attributes? I think baseIndexId and SelectedIndexId are enough

Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java:1992]: Why we should use selectedIndex as hash code？equals function contains it

**Comment by 924060929** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java (hunk 1 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java:1992]: if equals function contains selectedIndex, it is safe to remove selected index from hashCode function

**Review by morrySnow — APPROVED** [review]

**Review by starocean999 — APPROVED** [review]

**Pull request merged**